File: src/utils/data/MinMaxQueue.java
package utils.data;

import java.io.*;
import java.util.*;

import utils.constants.MathValueType;

public class MinMaxQueue<T> implements Iterable<T>, Serializable {
    private static final long serialVersionUID = 1L;
    
    private SortedListAvl<T> elements;
    
    private MathValueType oversizeOutType = MathValueType.MIN;
    
    private int queueSize = -1;
    
    public MinMaxQueue() {
        this(-1, MathValueType.MIN, new MinMaxQueue.DefaultComparator<T>());
    }
    
    public MinMaxQueue(Comparator<T> comparator) {
        this(-1, MathValueType.MIN, comparator);
    }
    
    public MinMaxQueue(int maxSize, MathValueType oversizeOutType) {
        this(maxSize, oversizeOutType, new DefaultComparator<T>());
    }
    
    public MinMaxQueue(int maxSize, MathValueType oversizeOutType, Comparator<T> comparator) {
        elements = new SortedListAvl<T>(comparator);
        queueSize = maxSize;
        setOversizeOutType(oversizeOutType);
    }
    
    public void add(T ... values) {
        for (T value : values) {
            add(value);
        }
    }
    
    public void add(T o) {
        if (queueSize > 0 && size() == queueSize) {
            elements.add(o);
            if (oversizeOutType == MathValueType.MIN) {
                elements.pollMin();
            } else {
                elements.pollMax();
            }
        } else {
            elements.add(o);
        }
    }
    
    public void addAll(MinMaxQueue<T> otherQueue) {
        if (otherQueue != null) {
            for (int i = 0; i < otherQueue.size(); i++) {
                add(otherQueue.get(i));
            }
        }
    }
    
    public void addAll(Collection<T> values) {
        for (T value : values) {
            add(value);
        }
    }
    
    public int size() {
        return elements.size();
    }
    
    public T pollMin() {
        return elements.pollMin();
    }
    
    public T pollMax() {
        return elements.pollMax();
    }
    
    public T peekMin() {
        return elements.getMin();
    }
    
    public T peekMax() {
        return elements.getMax();
    }
    
    public T peekIndex(int i) {
        if (i >= 0 && i < elements.size()) {
            return elements.get(i);
        } else {
            return null;
        }
    }
    
    public T getMin() {
        return peekMin();
    }
    
    public T getMax() {
        return peekMax();
    }
    
    public T get(int i) {
        return elements.get(i);
    }
    
    public boolean remove(T t) {
        return elements.remove(t);
    }
    
    public T remove(int i) {
        return removeAt(i);
    }

    public T removeAt(int i) {
        return elements.removeAt(i);
    }
    
    public int indexOf(T o) {
        return elements.indexOf(o);
    }
    
    public T removeFirst() {
        return pollMin();
    }
    
    public T removeMin() {
        return pollMin();
    }
    
    public T removeLast() {
        return pollMax();
    }
    
    public T removeMax() {
        return pollMax();
    }
    
    public void clear() {
        elements.clear();
    }
    
    public void setQueueSize(int qs) {
        queueSize = qs;
        if (queueSize > 0 && size() > queueSize) {
            while (size() > queueSize) {
                if (oversizeOutType == MathValueType.MIN) {
                    pollMin();
                } else {
                    pollMax();
                }
            }
        }
    }
    
    public void setOversizeOutType(MathValueType oversizeOutType) {
        if (oversizeOutType != null) {
            this.oversizeOutType = oversizeOutType;
        }
    }
    
    public int getQueueSize() {
        return queueSize;
    }
    
    public MathValueType getOversizeOutType() {
        return oversizeOutType;
    }
    
    public SortedListAvl<T> getList() {
        return elements.clone();
    }
    
    public void reorder() {
        elements.resort();
    }
    
    public boolean isEmpty() {
        return elements.isEmpty();
    }

    public Iterator<T> iterator() {
        return elements.listIterator();
    }
    
    private static class DefaultComparator<T> implements Comparator<T> {
        @SuppressWarnings({ "unchecked", "rawtypes" })
        public int compare(T t1, T t2) {
            if (t1 instanceof Comparable) {
                int result = ((Comparable)t1).compareTo(t2); 
                return result;
            } else {
                throw new NotComparableException("Not comparable exception");
            }
        }
    }
}

File: src/utils/data/MinMaxQueue.kt
package utils.data;

import java.io.*;
import java.util.*;

import utils.constants.MathValueType;

class MinMaxQueue<T>: Iterable<T>, Serializable {
    private val elements: SortedListAvl<T>
    
    var oversizeOutType = MathValueType.MIN
        private set
    
    var size: Int = -1
        set(value) {
            field = size
            if (field > 0 && elements.size > field) {
                while (elements.size > field) {
                    if (oversizeOutType == MathValueType.MIN) {
                        pollMin()
                    } else {
                        pollMax()
                    }
                }
            }
        }
        
    
    constructor(): this(-1, MathValueType.MIN, MinMaxQueue.DefaultComparator<T>())
    
    constructor(comparator: Comparator<T>): this(-1, MathValueType.MIN, comparator)
    
    constructor(maxSize: Int, oversizeOutType: MathValueType): this(maxSize, oversizeOutType, DefaultComparator<T>())
    
    constructor(maxSize: Int, oversizeOutType: MathValueType, comparator: Comparator<T>) {
        elements = SortedListAvl<T>(comparator)
        size = maxSize
        this.oversizeOutType = oversizeOutType
    }
    
    fun add(vararg values: T) {
        for (value in values) {
            add(value)
        }
    }
    
    fun add(element: T) {
        if (size > 0) {
            elements.add(element);
            if (oversizeOutType == MathValueType.MIN) {
                elements.pollMin();
            } else {
                elements.pollMax();
            }
        } else {
            elements.add(element);
        }
    }
    
    fun addAll(otherQueue: MinMaxQueue<T>) {
        for (i in 0 .. otherQueue.size - 1) {
            add(otherQueue.get(i))
        }
    }
    
    fun addAll(values: Collection<T>) {
        for (value: T in values) {
            add(value)
        }
    }
    
    fun pollMin(): T {
        return elements.pollMin()
    }
    
    fun pollMax(): T {
        return elements.pollMax()
    }
    
    fun peekMin(): T {
        return elements.getMin()
    }
    
    fun peekMax(): T {
        return elements.getMax()
    }
    
    fun peekIndex(index: Int): T? {
        if (index >= 0 && index < elements.size) {
            return elements.get(index)
        } else {
            return null
        }
    }
    
    fun getMin(): T {
        return peekMin()
    }
    
    fun getMax(): T {
        return peekMax()
    }
    
    fun get(index: Int): T {
        return elements.get(index)
    }
    
    fun remove(element: T): Boolean {
        return elements.remove(element)
    }
    
    fun removeAt(index: Int): T {
        return elements.removeAt(index);
    }
    
    fun indexOf(element: T): Int {
        return elements.indexOf(element)
    }
    
    fun removeFirst(): T {
        return pollMin()
    }
    
    fun removeMin(): T {
        return pollMin()
    }
    
    fun removeLast(): T {
        return pollMax()
    }
    
    fun removeMax(): T {
        return pollMax()
    }
    
    fun clear() {
        elements.clear()
    }
    
    fun getList(): SortedListAvl<T> {
        return elements.clone()
    }
    
    fun reorder() {
        elements.resort()
    }
    
    fun isEmpty(): Boolean {
        return elements.isEmpty()
    }
    
    fun isNotEmpty(): Boolean {
        return elements.isNotEmpty()
    }

    override fun iterator(): Iterator<T> {
        return elements.listIterator()
    }
    
    class DefaultComparator<T>: Comparator<T> {
        override fun compare(t1: T, t2: T): Int {
            if (t1 is Comparable<*>) {
                val result = (t1 as Comparable<T>).compareTo(t2); 
                return result
            } else {
                throw NotComparableException("Not comparable exception")
            }
        }
    }
}